File: backend/services/policy_service.py
"""
Policy service - Business logic for policy operations
"""
from typing import Optional, List, Tuple
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, func
from datetime import datetime
from fastapi import UploadFile
import os
import uuid
from models import Policy, PolicyAcknowledgment, User, UserRole
from schemas.policy_schemas import PolicyCreate, PolicyUpdate, PolicyResponse
from config import settings


class PolicyService:
    """Policy service class"""
    
    @staticmethod
    def create_policy(
        db: Session,
        policy_data: PolicyCreate,
        created_by_user: User
    ) -> Policy:
        """
        Create a new policy
        
        Args:
            db: Database session
            policy_data: Policy data from request
            created_by_user: User creating the policy
            
        Returns:
            Policy: Created policy object
        """
        # Create policy
        new_policy = Policy(
            title=policy_data.title,
            description=policy_data.description,
            content=policy_data.content,
            category=policy_data.category,
            version=policy_data.version,
            effective_date=policy_data.effective_date,
            review_date=policy_data.review_date,
            created_by=created_by_user.id,
            created_at=datetime.utcnow(),
            is_active=True
        )
        
        db.add(new_policy)
        db.commit()
        db.refresh(new_policy)
        
        return new_policy
    
    @staticmethod
    def get_policies(
        db: Session,
        current_user: User,
        include_inactive: bool = False,
        category: Optional[str] = None,
        skip: int = 0,
        limit: int = 100
    ) -> Tuple[List[Policy], int, int]:
        """
        Get all policies
        
        Args:
            db: Database session
            current_user: Current user
            include_inactive: Include inactive policies
            category: Filter by category
            skip: Number of records to skip (pagination)
            limit: Maximum number of records to return
            
        Returns:
            Tuple of (policies, total_count, active_count)
        """
        # Build query
        query = db.query(Policy)
        
        # Filter by active status (unless including inactive)
        if not include_inactive:
            query = query.filter(Policy.is_active == True)
        
        # Filter by category
        if category:
            query = query.filter(Policy.category == category)
        
        # Order by effective date (newest first)
        query = query.order_by(Policy.effective_date.desc())
        
        # Get total count
        total_count = query.count()
        
        # Get active count
        active_count = db.query(Policy).filter(Policy.is_active == True).count()
        
        # Apply pagination
        policies = query.offset(skip).limit(limit).all()
        
        return policies, total_count, active_count
    
    @staticmethod
    def get_policy_by_id(
        db: Session,
        policy_id: int
    ) -> Optional[Policy]:
        """
        Get policy by ID
        
        Args:
            db: Database session
            policy_id: Policy ID
            
        Returns:
            Policy or None if not found
        """
        return db.query(Policy).filter(Policy.id == policy_id).first()
    
    @staticmethod
    def update_policy(
        db: Session,
        policy_id: int,
        update_data: PolicyUpdate
    ) -> Optional[Policy]:
        """
        Update policy
        
        Args:
            db: Database session
            policy_id: Policy ID
            update_data: Update data
            
        Returns:
            Updated policy or None if not found
        """
        policy = db.query(Policy).filter(Policy.id == policy_id).first()
        
        if not policy:
            return None
        
        # Update fields that are provided
        update_dict = update_data.model_dump(exclude_unset=True)
        for field, value in update_dict.items():
            setattr(policy, field, value)
        
        policy.updated_at = datetime.utcnow()
        
        db.commit()
        db.refresh(policy)
        
        return policy
    
    @staticmethod
    def delete_policy(
        db: Session,
        policy_id: int,
        soft_delete: bool = True
    ) -> bool:
        """
        Delete policy (soft delete by default)
        
        Args:
            db: Database session
            policy_id: Policy ID
            soft_delete: If True, set is_active to False; if False, actually delete
            
        Returns:
            True if deleted, False if not found
        """
        policy = db.query(Policy).filter(Policy.id == policy_id).first()
        
        if not policy:
            return False
        
        if soft_delete:
            # Soft delete: set is_active to False
            policy.is_active = False
            policy.updated_at = datetime.utcnow()
            db.commit()
        else:
            # Hard delete: remove from database (also deletes file if exists)
            if policy.document_path and os.path.exists(policy.document_path):
                os.remove(policy.document_path)
            db.delete(policy)
            db.commit()
        
        return True
    
    @staticmethod
    async def upload_policy_document(
        db: Session,
        policy_id: int,
        file: UploadFile
    ) -> Tuple[Optional[Policy], Optional[str], Optional[int]]:
        """
        Upload policy document (PDF)
        
        Args:
            db: Database session
            policy_id: Policy ID
            file: Uploaded file
            
        Returns:
            Tuple of (policy, file_path, file_size) or (None, error_message, None)
        """
        # Get policy
        policy = db.query(Policy).filter(Policy.id == policy_id).first()
        
        if not policy:
            return None, "Policy not found", None
        
        # Validate file type
        if not file.filename.lower().endswith('.pdf'):
            return None, "Only PDF files are allowed", None
        
        # Create unique filename
        file_extension = os.path.splitext(file.filename)[1]
        unique_filename = f"{policy_id}_{uuid.uuid4().hex[:8]}_{file.filename}"
        file_path = os.path.join(settings.UPLOAD_DIR, "policies", unique_filename)
        
        # Read file content and check size
        file_content = await file.read()
        file_size = len(file_content)
        
        if file_size > settings.MAX_FILE_SIZE:
            return None, f"File size exceeds maximum allowed ({settings.MAX_FILE_SIZE_MB}MB)", None
        
        # Delete old file if exists
        if policy.document_path and os.path.exists(policy.document_path):
            os.remove(policy.document_path)
        
        # Save file
        with open(file_path, "wb") as f:
            f.write(file_content)
        
        # Update policy with file path
        policy.document_path = file_path
        policy.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(policy)
        
        # Auto-index policy for RAG (async, non-blocking)
        try:
            from ai_services.policy_rag_service import PolicyRAGService
            rag_service = PolicyRAGService()
            rag_service.index_policy_document(file_path, policy.title)
        except Exception as e:
            # Log error but don't fail the upload
            import logging
            logger = logging.getLogger(__name__)
            logger.warning(f"Failed to index policy {policy.id} for RAG: {e}")
        
        return policy, file_path, file_size
    
    @staticmethod
    def get_policy_document_path(
        db: Session,
        policy_id: int
    ) -> Optional[str]:
        """
        Get policy document file path
        
        Args:
            db: Database session
            policy_id: Policy ID
            
        Returns:
            File path or None if not found
        """
        policy = db.query(Policy).filter(Policy.id == policy_id).first()
        
        if not policy or not policy.document_path:
            return None
        
        if not os.path.exists(policy.document_path):
            return None
        
        return policy.document_path
    
    @staticmethod
    def acknowledge_policy(
        db: Session,
        policy_id: int,
        user: User
    ) -> Optional[PolicyAcknowledgment]:
        """
        Acknowledge policy by user
        
        Args:
            db: Database session
            policy_id: Policy ID
            user: User acknowledging the policy
            
        Returns:
            PolicyAcknowledgment or None if already acknowledged
        """
        # Check if policy exists
        policy = db.query(Policy).filter(Policy.id == policy_id).first()
        if not policy:
            return None
        
        # Check if already acknowledged
        existing = db.query(PolicyAcknowledgment).filter(
            and_(
                PolicyAcknowledgment.policy_id == policy_id,
                PolicyAcknowledgment.user_id == user.id
            )
        ).first()
        
        if existing:
            return existing  # Already acknowledged
        
        # Create acknowledgment
        acknowledgment = PolicyAcknowledgment(
            policy_id=policy_id,
            user_id=user.id,
            acknowledged_date=datetime.utcnow()
        )
        
        db.add(acknowledgment)
        db.commit()
        db.refresh(acknowledgment)
        
        return acknowledgment
    
    @staticmethod
    def get_policy_acknowledgments(
        db: Session,
        policy_id: int,
        skip: int = 0,
        limit: int = 100
    ) -> Tuple[List[PolicyAcknowledgment], int]:
        """
        Get all acknowledgments for a policy
        
        Args:
            db: Database session
            policy_id: Policy ID
            skip: Pagination offset
            limit: Maximum results
            
        Returns:
            Tuple of (acknowledgments, total_count)
        """
        query = db.query(PolicyAcknowledgment).filter(
            PolicyAcknowledgment.policy_id == policy_id
        ).order_by(PolicyAcknowledgment.acknowledged_date.desc())
        
        total_count = query.count()
        acknowledgments = query.offset(skip).limit(limit).all()
        
        return acknowledgments, total_count
    
    @staticmethod
    def get_user_acknowledgments(
        db: Session,
        user_id: int
    ) -> List[PolicyAcknowledgment]:
        """
        Get all policies acknowledged by a user
        
        Args:
            db: Database session
            user_id: User ID
            
        Returns:
            List of acknowledgments
        """
        return db.query(PolicyAcknowledgment).filter(
            PolicyAcknowledgment.user_id == user_id
        ).order_by(PolicyAcknowledgment.acknowledged_date.desc()).all()
    
    @staticmethod
    def has_user_acknowledged_policy(
        db: Session,
        policy_id: int,
        user_id: int
    ) -> bool:
        """
        Check if user has acknowledged a policy
        
        Args:
            db: Database session
            policy_id: Policy ID
            user_id: User ID
            
        Returns:
            True if acknowledged, False otherwise
        """
        count = db.query(PolicyAcknowledgment).filter(
            and_(
                PolicyAcknowledgment.policy_id == policy_id,
                PolicyAcknowledgment.user_id == user_id
            )
        ).count()
        
        return count > 0
    
    @staticmethod
    def format_policy_response(
        policy: Policy,
        db: Session,
        current_user: Optional[User] = None
    ) -> PolicyResponse:
        """
        Format policy for response with additional fields
        
        Args:
            policy: Policy object
            db: Database session (to fetch creator name, acknowledgments)
            current_user: Current user (to check acknowledgment status)
            
        Returns:
            PolicyResponse with formatted data
        """
        # Get creator name
        creator = db.query(User).filter(User.id == policy.created_by).first()
        creator_name = creator.name if creator else "Unknown"
        
        # Check if has document
        has_document = bool(policy.document_path and os.path.exists(policy.document_path))
        
        # Generate document URL
        document_url = None
        if has_document:
            # Convert absolute path to relative URL
            document_url = f"/api/v1/policies/{policy.id}/download"
        
        # Get acknowledgment count
        acknowledgment_count = db.query(PolicyAcknowledgment).filter(
            PolicyAcknowledgment.policy_id == policy.id
        ).count()
        
        # Check if current user has acknowledged
        is_acknowledged_by_user = False
        if current_user:
            is_acknowledged_by_user = PolicyService.has_user_acknowledged_policy(
                db, policy.id, current_user.id
            )
        
        return PolicyResponse(
            id=policy.id,
            title=policy.title,
            description=policy.description,
            content=policy.content,
            category=policy.category,
            version=policy.version,
            is_active=policy.is_active,
            effective_date=policy.effective_date,
            review_date=policy.review_date,
            document_path=policy.document_path,
            created_by=policy.created_by,
            created_by_name=creator_name,
            created_at=policy.created_at,
            updated_at=policy.updated_at,
            has_document=has_document,
            document_url=document_url,
            acknowledgment_count=acknowledgment_count,
            is_acknowledged_by_user=is_acknowledged_by_user
        )
    
    @staticmethod
    def get_stats(db: Session) -> dict:
        """
        Get policy statistics
        
        Args:
            db: Database session
            
        Returns:
            Dictionary with statistics
        """
        total = db.query(Policy).count()
        active = db.query(Policy).filter(Policy.is_active == True).count()
        inactive = db.query(Policy).filter(Policy.is_active == False).count()
        
        # Policies with documents
        with_documents = db.query(Policy).filter(
            and_(
                Policy.is_active == True,
                Policy.document_path != None
            )
        ).count()
        
        # Total acknowledgments
        total_acknowledgments = db.query(PolicyAcknowledgment).count()
        
        # Get category breakdown
        categories = db.query(
            Policy.category,
            func.count(Policy.id).label('count')
        ).filter(
            Policy.is_active == True
        ).group_by(Policy.category).all()
        
        category_stats = {cat: count for cat, count in categories if cat}
        
        return {
            "total": total,
            "active": active,
            "inactive": inactive,
            "with_documents": with_documents,
            "total_acknowledgments": total_acknowledgments,
            "categories": category_stats
        }

File: frontend/src/services/policyService.ts
/**
 * Policy Service
 * Service layer for policies API calls with file upload/download support
 */

import api from './api';

// Types
export interface Policy {
  id: number;
  title: string;
  description?: string;
  content: string;
  category?: string;
  version: string;
  is_active: boolean;
  effective_date: string;
  review_date?: string;
  document_path?: string;
  created_by: number;
  created_at: string;
  updated_at: string;
  created_by_name?: string;
  has_document: boolean;
  document_url?: string;
  acknowledgment_count: number;
  is_acknowledged_by_user: boolean;
}

export interface PolicyListResponse {
  policies: Policy[];
  total: number;
  active: number;
}

export interface CreatePolicyData {
  title: string;
  description?: string;
  content: string;
  category?: string;
  version?: string;
  effective_date: string;
  review_date?: string;
  require_acknowledgment?: boolean;
}

export interface UpdatePolicyData {
  title?: string;
  description?: string;
  content?: string;
  category?: string;
  version?: string;
  effective_date?: string;
  review_date?: string;
  is_active?: boolean;
}

export interface PolicyAcknowledgment {
  id: number;
  policy_id: number;
  user_id: number;
  acknowledged_date: string;
  policy_title?: string;
  user_name?: string;
}

export interface PolicyAcknowledgmentListResponse {
  acknowledgments: PolicyAcknowledgment[];
  total: number;
}

export interface PolicyUploadResponse {
  policy_id: number;
  file_name: string;
  file_path: string;
  file_size: number;
  message: string;
}

export interface PolicyStats {
  total: number;
  active: number;
  inactive: number;
  with_documents: number;
  total_acknowledgments: number;
  categories: Record<string, number>;
}

/**
 * Policy Service Class
 */
class PolicyService {
  private baseUrl = '/policies';

  /**
   * Get all policies
   * @param params Query parameters for filtering and pagination
   * @returns List of policies with metadata
   */
  async getPolicies(params?: {
    include_inactive?: boolean;
    category?: string;
    skip?: number;
    limit?: number;
  }): Promise<PolicyListResponse> {
    try {
      const response = await api.get<PolicyListResponse>(this.baseUrl, {
        params: params || {},
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching policies:', error);
      throw error;
    }
  }

  /**
   * Get policy by ID
   * @param id Policy ID
   * @returns Policy object
   */
  async getPolicyById(id: number): Promise<Policy> {
    try {
      const response = await api.get<Policy>(`${this.baseUrl}/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching policy ${id}:`, error);
      throw error;
    }
  }

  /**
   * Create a new policy (HR only)
   * @param data Policy data
   * @returns Created policy
   */
  async createPolicy(data: CreatePolicyData): Promise<Policy> {
    try {
      const response = await api.post<Policy>(this.baseUrl, data);
      return response.data;
    } catch (error) {
      console.error('Error creating policy:', error);
      throw error;
    }
  }

  /**
   * Update a policy (HR only)
   * @param id Policy ID
   * @param data Update data
   * @returns Updated policy
   */
  async updatePolicy(
    id: number,
    data: UpdatePolicyData
  ): Promise<Policy> {
    try {
      const response = await api.put<Policy>(`${this.baseUrl}/${id}`, data);
      return response.data;
    } catch (error) {
      console.error(`Error updating policy ${id}:`, error);
      throw error;
    }
  }

  /**
   * Delete a policy (HR only)
   * @param id Policy ID
   * @param hardDelete Whether to permanently delete (default: false)
   * @returns Success message
   */
  async deletePolicy(
    id: number,
    hardDelete: boolean = false
  ): Promise<{ message: string }> {
    try {
      const response = await api.delete<{ message: string }>(
        `${this.baseUrl}/${id}`,
        {
          params: { hard_delete: hardDelete },
        }
      );
      return response.data;
    } catch (error) {
      console.error(`Error deleting policy ${id}:`, error);
      throw error;
    }
  }

  /**
   * Upload policy document (PDF) (HR only)
   * @param policyId Policy ID
   * @param file PDF file to upload
   * @returns Upload response with file details
   */
  async uploadPolicyDocument(
    policyId: number,
    file: File
  ): Promise<PolicyUploadResponse> {
    try {
      // Create form data
      const formData = new FormData();
      formData.append('file', file);

      // Upload with multipart/form-data
      const response = await api.post<PolicyUploadResponse>(
        `${this.baseUrl}/${policyId}/upload`,
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error(`Error uploading policy document for ${policyId}:`, error);
      throw error;
    }
  }

  /**
   * Download policy document (PDF)
   * @param policyId Policy ID
   * @returns Blob of the PDF file
   */
  async downloadPolicyDocument(policyId: number): Promise<Blob> {
    try {
      const response = await api.get<Blob>(
        `${this.baseUrl}/${policyId}/download`,
        {
          responseType: 'blob',
        }
      );
      return response.data;
    } catch (error) {
      console.error(`Error downloading policy document ${policyId}:`, error);
      throw error;
    }
  }

  /**
   * Download policy document and trigger browser download
   * @param policyId Policy ID
   * @param filename Custom filename (optional)
   */
  async downloadAndSavePolicyDocument(
    policyId: number,
    filename?: string
  ): Promise<void> {
    try {
      const blob = await this.downloadPolicyDocument(policyId);

      // Create download link
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement('a');
      link.href = url;
      link.download = filename || `policy-${policyId}.pdf`;
      document.body.appendChild(link);
      link.click();

      // Cleanup
      document.body.removeChild(link);
      window.URL.revokeObjectURL(url);
    } catch (error) {
      console.error('Error triggering policy download:', error);
      throw error;
    }
  }

  /**
   * Acknowledge a policy
   * @param policyId Policy ID
   * @returns Acknowledgment record
   */
  async acknowledgePolicy(policyId: number): Promise<PolicyAcknowledgment> {
    try {
      const response = await api.post<PolicyAcknowledgment>(
        `${this.baseUrl}/${policyId}/acknowledge`
      );
      return response.data;
    } catch (error) {
      console.error(`Error acknowledging policy ${policyId}:`, error);
      throw error;
    }
  }

  /**
   * Get policy acknowledgments (HR only)
   * @param policyId Policy ID
   * @param params Pagination parameters
   * @returns List of acknowledgments
   */
  async getPolicyAcknowledgments(
    policyId: number,
    params?: {
      skip?: number;
      limit?: number;
    }
  ): Promise<PolicyAcknowledgmentListResponse> {
    try {
      const response = await api.get<PolicyAcknowledgmentListResponse>(
        `${this.baseUrl}/${policyId}/acknowledgments`,
        {
          params: params || {},
        }
      );
      return response.data;
    } catch (error) {
      console.error(`Error fetching policy acknowledgments for ${policyId}:`, error);
      throw error;
    }
  }

  /**
   * Get policy statistics (HR only)
   * @returns Statistics object
   */
  async getStatistics(): Promise<PolicyStats> {
    try {
      const response = await api.get<PolicyStats>(
        `${this.baseUrl}/stats/summary`
      );
      return response.data;
    } catch (error) {
      console.error('Error fetching policy statistics:', error);
      throw error;
    }
  }

  /**
   * Get active policies (helper method)
   * @param limit Number of policies to fetch (default: 10)
   * @returns Active policies
   */
  async getActivePolicies(limit: number = 10): Promise<Policy[]> {
    try {
      const response = await this.getPolicies({ limit, include_inactive: false });
      return response.policies;
    } catch (error) {
      console.error('Error fetching active policies:', error);
      throw error;
    }
  }

  /**
   * Get policies by category
   * @param category Category name
   * @param limit Maximum results
   * @returns Policies in category
   */
  async getPoliciesByCategory(category: string, limit: number = 100): Promise<Policy[]> {
    try {
      const response = await this.getPolicies({ category, limit });
      return response.policies;
    } catch (error) {
      console.error(`Error fetching policies for category ${category}:`, error);
      throw error;
    }
  }

  /**
   * Check if policy requires acknowledgment
   * @param policy Policy object
   * @returns True if user hasn't acknowledged yet
   */
  requiresAcknowledgment(policy: Policy): boolean {
    return !policy.is_acknowledged_by_user;
  }

  /**
   * Format policy date for display
   * @param dateString ISO date string
   * @returns Formatted date string
   */
  formatDate(dateString: string): string {
    const date = new Date(dateString);
    return date.toLocaleDateString('en-IN', {
      day: '2-digit',
      month: 'short',
      year: 'numeric',
    });
  }

  /**
   * Format policy datetime for display
   * @param dateString ISO datetime string
   * @returns Formatted datetime string
   */
  formatDateTime(dateString: string): string {
    const date = new Date(dateString);
    return date.toLocaleString('en-IN', {
      day: '2-digit',
      month: 'short',
      year: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    });
  }

  /**
   * Get file size in human-readable format
   * @param bytes File size in bytes
   * @returns Formatted file size
   */
  formatFileSize(bytes: number): string {
    if (bytes === 0) return '0 Bytes';
    const k = 1024;
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return Math.round((bytes / Math.pow(k, i)) * 100) / 100 + ' ' + sizes[i];
  }

  /**
   * Validate PDF file before upload
   * @param file File to validate
   * @param maxSizeMB Maximum file size in MB
   * @returns Error message if invalid, null if valid
   */
  validatePolicyFile(file: File, maxSizeMB: number = 10): string | null {
    // Check file type
    if (file.type !== 'application/pdf' && !file.name.toLowerCase().endsWith('.pdf')) {
      return 'Only PDF files are allowed';
    }

    // Check file size
    const maxSizeBytes = maxSizeMB * 1024 * 1024;
    if (file.size > maxSizeBytes) {
      return `File size exceeds maximum allowed (${maxSizeMB}MB)`;
    }

    // Check filename
    if (file.name.length > 255) {
      return 'Filename is too long';
    }

    return null;
  }
}

// Export singleton instance
const policyService = new PolicyService();
export default policyService;

